src/client/datascience/kernel-launcher/localKernelSpecFinderBase.ts

```diff
@@ -126,19 +137,42 @@ export abstract class LocalKernelSpecFinderBase {
         if (!this.pathToKernelSpec.has(specPath)) {
             this.pathToKernelSpec.set(specPath, this.loadKernelSpec(specPath, interpreter, cancelToken));
         }
-
         // ! as the has and set above verify that we have a return here
-        return this.pathToKernelSpec.get(specPath)!.then((value) => {
-            if (value) {
-                return value;
+        return this.pathToKernelSpec.get(specPath)!.then((kernelSpec) => {
+            // Delete old kernelSpecs that we created in the global kernelSpecs folder.
+            const shouldDeleteKernelSpec =
+                kernelSpec &&
+                globalSpecRootPath &&
+                getKernelRegistrationInfo(kernelSpec) &&
+                kernelSpec.specFile &&
+                path.dirname(path.dirname(kernelSpec.specFile)) === globalSpecRootPath;
+            if (kernelSpec && !shouldDeleteKernelSpec) {
+                return kernelSpec;
             }
-            // If we failed to get a kernelspec full path from our cache and loaded list
+            if (kernelSpec?.specFile && shouldDeleteKernelSpec) {
+                // If this kernelSpec was registered by us and is in the global kernels folder,
+                // then remove it.
+                this.deleteOldKernelSpec(kernelSpec.specFile).catch(noop);
```

Comment:
This seems like unexpected behavior - the result of asking for a kernelSpec that exists, is for it to be deleted

Reply:
Yes, its a side effect, we need to delete the old kernelspecs & this is a one time task.
I don't want to create a whole new method to iterate through kernels all over again (which is disc io & unnecessary complication).
Also I need a place to filter out these old kernlespecs (hence this is the spot).
In the future this code can be removed.